Modules/timemodule.c

```diff
@@ -2106,12 +2107,60 @@ pysleep(_PyTime_t secs)
             PyErr_SetFromErrno(PyExc_OSError);
             return -1;
         }
+
+        /* sleep was interrupted by SIGINT */
+        if (PyErr_CheckSignals()) {
+            return -1;
+        }
+
+#ifndef HAVE_CLOCK_NANOSLEEP
+        if (get_monotonic(&monotonic) < 0) {
+            return -1;
+        }
+        secs = deadline - monotonic;
+        if (secs < 0) {
+            ret = 0;
+            break;
+        }
+        /* retry with the recomputed delay */
+#endif
+    } while (1);
 #else
+    int is_timer_ready = 0;
+    _PyTime_t millisecs;
+    DWORD ul_millis;
+    DWORD rc;
+    FILETIME FileTime;
+    LARGE_INTEGER deadline;
+    /* hTimer, hInterruptEvent */
+    HANDLE hEvents[EVENT_SIZE] = {NULL, NULL};
+
+    GetSystemTimePreciseAsFileTime(&FileTime);
+    CopyMemory(&deadline, &FileTime, sizeof(FILETIME));
+
+    /* convert to 100 nsec unit and add */
+    deadline.QuadPart += _PyTime_As100Nanoseconds(secs, _PyTime_ROUND_CEILING);
+
+    hEvents[EVENT_TIMER] = CreateWaitableTimerW(NULL, FALSE, NULL);
+    if (NULL == hEvents[EVENT_TIMER]) {
+        return -1;
+    }
+
+    if (!SetWaitableTimer(hEvents[EVENT_TIMER], &deadline, 0, NULL, NULL, FALSE)) {
+        ret = -1;
+        is_timer_ready = 0;
+    }
+    else {
+        is_timer_ready = 1;
+    }
+
+    while (is_timer_ready) {
         millisecs = _PyTime_AsMilliseconds(secs, _PyTime_ROUND_CEILING);
-        if (millisecs > (double)ULONG_MAX) {
+        if (millisecs > LLONG_MAX) {
```

Comment:
In Windows, as is see in Visual Studio 2019 there is a conflict and warning in this comparing.

- Type of millisecs is `_PyTime_t (int64_t)`, see in [pytime.h](https://github.com/python/cpython/blob/main/Include/cpython/pytime.h#L16).
- Comparing between int64_t and double is not too nice.
- My code is also not correct, because as i see it needs to check the overflow situation of ul_millis which will be `unsigned long`, later in code.

I will fix for the following:
```C
        if (millisecs > ((_PyTime_t)ULONG_MAX)) {
            PyErr_SetString(PyExc_OverflowError,
                "sleep length is too large");
            ret = -1;
            break;
        }
```

Reply:
I'm quite sure that old code used the `double` type and was modified to use the `_PyTime_t` type. The `(double)` was properly missed: it should be removed.

`_PyType_t` is signed, whereas `unsigned long` is not. I don't know what happens if sizeof(_PyTime_t) == sizeof(unsigned long): when long size is 64-bit, which is the case on 64-bit Linux for example. Maybe the whole test (if ... { ... }) should be skipped if sizeof(unsigned long) == sizeof(_PyTime_t).